# RNA-seq processing

*This README will be finished soon* 

Here, you can find a detailed description of how RNA-seq data was processed in this project.

## Dependencies

* [Trim Galore](https://github.com/FelixKrueger/TrimGalore)
* [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
* [STAR](https://github.com/alexdobin/STAR/tree/master)
* [samtools](http://www.htslib.org/download/)
* [sambamba](https://github.com/biod/sambamba)
* [featureCounts]([https://github.com/macs3-project/MACS](https://rnnh.github.io/bioinfo-notebook/docs/featureCounts.html))
* [DESeq2 (R package)](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)


## Summary of the workflow

1. **Trimming and Quality**: Trim Galore and FastQC
2. **Alignment**: Bowtie2
3. **Filtering**: samtools and sambamba
5. **Counting**: featureCounts
6. **Differential Analysis**: DESeq2


## 1. Trimming and Quality

## 2. Alignment

## 3. Filtering

## 4. Counting

## 5. Differential Analysis


